// web-audio-engine/lib/impl/BasePannerNode.js
"use strict";

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var AudioNode = require("./AudioNode");

var _require = require("../utils"),
    defaults = _require.defaults;

var _require2 = require("../utils"),
    toNumber = _require2.toNumber;

var _require3 = require("../constants/ChannelCountMode"),
    CLAMPED_MAX = _require3.CLAMPED_MAX,
    EXPLICIT = _require3.EXPLICIT;

var PanningModelTypes = ["equalpower", "HRTF"];
var DistanceModelTypes = ["linear", "inverse", "exponential"];

var DEFAULT_PANNING_MODEL = "equalpower";
var DEFAULT_DISTANCE_MODEL = "inverse";
var DEFAULT_REF_DISTANCE = 1;
var DEFAULT_MAX_DISTANCE = 10000;
var DEFAULT_ROLLOFF_FACTOR = 1;
var DEFAULT_CONE_INNER_ANGLE = 360;
var DEFAULT_CONE_OUTER_ANGLE = 360;
var DEFAULT_CONE_OUTER_GAIN = 0;

var BasePannerNode = function (_AudioNode) {
  _inherits(BasePannerNode, _AudioNode);

  /**
   * @param {AudioContext} context
   * @param {object}       opts
   * @param {string}       opts.panningModel
   * @param {string}       opts.distanceModel
   * @param {number}       opts.refDistance
   * @param {number}       opts.maxDistance
   * @param {number}       opts.rolloffFactor
   * @param {number}       opts.coneInnerAngle
   * @param {number}       opts.coneOuterAngle
   * @param {number}       opts.coneOuterGain
   */
  function BasePannerNode(context) {
    var opts = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {};

    _classCallCheck(this, BasePannerNode);

    var panningModel = defaults(opts.panningModel, DEFAULT_PANNING_MODEL);
    var distanceModel = defaults(opts.distanceModel, DEFAULT_DISTANCE_MODEL);
    var refDistance = defaults(opts.refDistance, DEFAULT_REF_DISTANCE);
    var maxDistance = defaults(opts.maxDistance, DEFAULT_MAX_DISTANCE);
    var rolloffFactor = defaults(opts.rolloffFactor, DEFAULT_ROLLOFF_FACTOR);
    var coneInnerAngle = defaults(opts.coneInnerAngle, DEFAULT_CONE_INNER_ANGLE);
    var coneOuterAngle = defaults(opts.coneOuterAngle, DEFAULT_CONE_OUTER_ANGLE);
    var coneOuterGain = defaults(opts.coneOuterGain, DEFAULT_CONE_OUTER_GAIN);

    var _this = _possibleConstructorReturn(this, (BasePannerNode.__proto__ || Object.getPrototypeOf(BasePannerNode)).call(this, context, opts, {
      inputs: [1],
      outputs: [2],
      channelCount: 2,
      channelCountMode: CLAMPED_MAX,
      allowedMaxChannelCount: 2,
      allowedChannelCountMode: [CLAMPED_MAX, EXPLICIT]
    }));

    _this._panningModel = panningModel;
    _this._distanceModel = distanceModel;
    _this._refDistance = refDistance;
    _this._maxDistance = maxDistance;
    _this._rolloffFactor = rolloffFactor;
    _this._coneInnerAngle = coneInnerAngle;
    _this._coneOuterAngle = coneOuterAngle;
    _this._coneOuterGain = coneOuterGain;
    return _this;
  }

  /**
   * @return {string}
   */


  _createClass(BasePannerNode, [{
    key: "getPanningModel",
    value: function getPanningModel() {
      return this._panningModel;
    }

    /**
     * @param {string} value
     */

  }, {
    key: "setPanningModel",
    value: function setPanningModel(value) {
      /* istanbul ignore else */
      if (PanningModelTypes.indexOf(value) !== -1) {
        this._panningModel = value;
      }
    }

    /**
     * @return {string}
     */

  }, {
    key: "getDistanceModel",
    value: function getDistanceModel() {
      return this._distanceModel;
    }

    /**
     * @param {string} value
     */

  }, {
    key: "setDistanceModel",
    value: function setDistanceModel(value) {
      /* istanbul ignore else */
      if (DistanceModelTypes.indexOf(value) !== -1) {
        this._distanceModel = value;
      }
    }

    /**
     * @return {number}
     */

  }, {
    key: "getRefDistance",
    value: function getRefDistance() {
      return this._refDistance;
    }

    /**
     * @param {number} value
     */

  }, {
    key: "setRefDistance",
    value: function setRefDistance(value) {
      this._refDistance = toNumber(value);
    }

    /**
     * @return {number}
     */

  }, {
    key: "getMaxDistance",
    value: function getMaxDistance() {
      return this._maxDistance;
    }

    /**
     * @param {number} value
     */

  }, {
    key: "setMaxDistance",
    value: function setMaxDistance(value) {
      this._maxDistance = toNumber(value);
    }

    /**
     * @return {number}
     */

  }, {
    key: "getRolloffFactor",
    value: function getRolloffFactor() {
      return this._rolloffFactor;
    }

    /**
     * @param {number} value
     */

  }, {
    key: "setRolloffFactor",
    value: function setRolloffFactor(value) {
      this._rolloffFactor = toNumber(value);
    }

    /**
     * @return {number}
     */

  }, {
    key: "getConeInnerAngle",
    value: function getConeInnerAngle() {
      return this._coneInnerAngle;
    }

    /**
     * @param {number} value
     */

  }, {
    key: "setConeInnerAngle",
    value: function setConeInnerAngle(value) {
      this._coneInnerAngle = toNumber(value);
    }

    /**
     * @return {number}
     */

  }, {
    key: "getConeOuterAngle",
    value: function getConeOuterAngle() {
      return this._coneOuterAngle;
    }

    /**
     * @param {number} value
     */

  }, {
    key: "setConeOuterAngle",
    value: function setConeOuterAngle(value) {
      this._coneOuterAngle = toNumber(value);
    }

    /**
     * @return {number}
     */

  }, {
    key: "getConeOuterGain",
    value: function getConeOuterGain() {
      return this._coneOuterGain;
    }

    /**
     * @param {number} value
     */

  }, {
    key: "setConeOuterGain",
    value: function setConeOuterGain(value) {
      this._coneOuterGain = toNumber(value);
    }
  }]);

  return BasePannerNode;
}(AudioNode);

module.exports = BasePannerNode;
