// web-audio-engine/lib/impl/ConstantSourceNode.js
"use strict";

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var AudioScheduledSourceNode = require("./AudioScheduledSourceNode");
var ConstantSourceNodeDSP = require("./dsp/ConstantSourceNode");

var _require = require("../utils"),
    defaults = _require.defaults;

var _require2 = require("../constants/ChannelCountMode"),
    MAX = _require2.MAX;

var _require3 = require("../constants/AudioParamRate"),
    AUDIO_RATE = _require3.AUDIO_RATE;

var DEFAULT_OFFSET = 1;

var ConstantSourceNode = function (_AudioScheduledSource) {
  _inherits(ConstantSourceNode, _AudioScheduledSource);

  /**
   * @param {AudioContext} context
   * @param {object}       opts
   * @param {number}       opts.offset
   */
  function ConstantSourceNode(context) {
    var opts = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {};

    _classCallCheck(this, ConstantSourceNode);

    var offset = defaults(opts.offset, DEFAULT_OFFSET);

    var _this = _possibleConstructorReturn(this, (ConstantSourceNode.__proto__ || Object.getPrototypeOf(ConstantSourceNode)).call(this, context, opts, {
      inputs: [1],
      outputs: [1],
      channelCount: 2,
      channelCountMode: MAX
    }));

    _this._offset = _this.addParam(AUDIO_RATE, offset);
    return _this;
  }

  /**
   * @return {AudioParam}
   */


  _createClass(ConstantSourceNode, [{
    key: "getOffset",
    value: function getOffset() {
      return this._offset;
    }
  }]);

  return ConstantSourceNode;
}(AudioScheduledSourceNode);

Object.assign(ConstantSourceNode.prototype, ConstantSourceNodeDSP);

module.exports = ConstantSourceNode;
