// web-audio-engine/lib/impl/DynamicsCompressorNode.js
"use strict";

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var AudioNode = require("./AudioNode");
var DynamicsCompressorNodeDSP = require("./dsp/DynamicsCompressorNode");

var _require = require("../utils"),
    defaults = _require.defaults;

var _require2 = require("../constants/ChannelCountMode"),
    EXPLICIT = _require2.EXPLICIT;

var _require3 = require("../constants/AudioParamRate"),
    CONTROL_RATE = _require3.CONTROL_RATE;

var DEFAULT_THRESHOLD = -24;
var DEFAULT_KNEE = 30;
var DEFAULT_RATIO = 12;
var DEFAULT_ATTACK = 0.003;
var DEFAULT_RELEASE = 0.25;

var DynamicsCompressorNode = function (_AudioNode) {
  _inherits(DynamicsCompressorNode, _AudioNode);

  /**
   * @param {AudioContext} context
   * @param {object}       opts
   * @param {number}       opts.threshold
   * @param {number}       opts.knee
   * @param {number}       opts.ratio
   * @param {number}       opts.attack
   * @param {number}       opts.release
   */
  function DynamicsCompressorNode(context) {
    var opts = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {};

    _classCallCheck(this, DynamicsCompressorNode);

    var threshold = defaults(opts.threshold, DEFAULT_THRESHOLD);
    var knee = defaults(opts.knee, DEFAULT_KNEE);
    var ratio = defaults(opts.ratio, DEFAULT_RATIO);
    var attack = defaults(opts.attack, DEFAULT_ATTACK);
    var release = defaults(opts.release, DEFAULT_RELEASE);

    var _this = _possibleConstructorReturn(this, (DynamicsCompressorNode.__proto__ || Object.getPrototypeOf(DynamicsCompressorNode)).call(this, context, opts, {
      inputs: [1],
      outputs: [2],
      channelCount: 2,
      channelCountMode: EXPLICIT
    }));

    _this._threshold = _this.addParam(CONTROL_RATE, threshold);
    _this._knee = _this.addParam(CONTROL_RATE, knee);
    _this._ratio = _this.addParam(CONTROL_RATE, ratio);
    _this._attack = _this.addParam(CONTROL_RATE, attack);
    _this._release = _this.addParam(CONTROL_RATE, release);
    return _this;
  }

  /**
   * @param {AudioParam}
   */


  _createClass(DynamicsCompressorNode, [{
    key: "getThreshold",
    value: function getThreshold() {
      return this._threshold;
    }

    /**
     * @param {AudioParam}
     */

  }, {
    key: "getKnee",
    value: function getKnee() {
      return this._knee;
    }

    /**
     * @param {AudioParam}
     */

  }, {
    key: "getRatio",
    value: function getRatio() {
      return this._ratio;
    }

    /**
     * @return {number}
     */
    /* istanbul ignore next */

  }, {
    key: "getReduction",
    value: function getReduction() {
      throw new TypeError("NOT YET IMPLEMENTED");
    }

    /**
     * @param {AudioParam}
     */

  }, {
    key: "getAttack",
    value: function getAttack() {
      return this._attack;
    }

    /**
     * @param {AudioParam}
     */

  }, {
    key: "getRelease",
    value: function getRelease() {
      return this._release;
    }
  }]);

  return DynamicsCompressorNode;
}(AudioNode);

Object.assign(DynamicsCompressorNode.prototype, DynamicsCompressorNodeDSP);

module.exports = DynamicsCompressorNode;
